File: src/types/analysis.ts
// =====================================================================
// src/types/analysis.ts
// 역할: 백엔드 API 응답과 프론트엔드 컴포넌트 사이에서 공유하는 타입 정의
// - backend/schemas/analysis.py 의 Pydantic 모델과 구조가 일치해야 함
// - 미정 판단이 필요한 경우 isUnspecified() 유틸 함수 사용
// =====================================================================

// ── 결정사항 ─────────────────────────────────────────────────────────
export interface Decision {
  content: string;        // 결정 내용
  evidence_text: string;  // 원문 근거
}

// ── 실행 항목 ───────────────────────────────────────────────────────
export interface ActionItem {
  task: string;               // 해야 할 일
  owner: string | null;       // 담당자 (null 이면 미정)
  due_date: string | null;    // 마감일 (null 이면 미정)
  priority: string | null;    // 우선순위: "높음" | "보통" | "낮음" | null
  evidence_text: string;      // 원문 근거
}

// ── 리스크 ────────────────────────────────────────────────────────────
export interface Risk {
  risk: string;               // 리스크 내용
  response: string | null;    // 대응 방향 (null 이면 미정)
  evidence_text: string;      // 원문 근거
}

// ── 분석 결과 전체 ────────────────────────────────────────────────────
export interface AnalysisResult {
  summary: string;
  decisions: Decision[];
  action_items: ActionItem[];
  risks: Risk[];
  missing_info: string[];
}

// ── 저장 API 요청/응답 ───────────────────────────────────────────────
export interface SaveGoogleDocRequest {
  title: string;           // 빈 문자열이면 백엔드에서 자동 생성
  result: AnalysisResult;
}

export interface SaveGoogleDocResponse {
  title: string;           // 실제로 저장된 문서명
  url: string;             // Google Docs URL
}

// ── 미정 판단 유틸 ────────────────────────────────────────────────────
/**
 * 역할: 값이 "미정"으로 간주되는지 판단한다.
 * - null, undefined, 빈 문자열, 또는 아래 키워드는 모두 미정으로 처리
 * - UI에서 담당자/마감일 등을 '미정' 뱃지로 표시할 때 사용
 */
const UNSPECIFIED_VALUES = new Set([
  "미정",
  "",
  "추후 결정",
  "미확정",
  "불명확",
  "없음",
]);

export function isUnspecified(value: string | null | undefined): boolean {
  if (value === null || value === undefined) return true;
  return UNSPECIFIED_VALUES.has(value.trim());
}

// ── 메트릭 계산 유틸 ─────────────────────────────────────────────────
/**
 * 역할: 분석 결과에서 대시보드용 통계 수치를 계산한다.
 * - App.tsx 의 Stats Cards 섹션에서 사용
 */
export interface AnalysisMetrics {
  totalActionItems: number;   // 실행 항목 수
  unassignedCount: number;    // 담당자 미정 수
  noDueDateCount: number;     // 마감일 미정 수
  riskCount: number;          // 리스크 수
}

export function calcMetrics(result: AnalysisResult): AnalysisMetrics {
  return {
    totalActionItems: result.action_items.length,
    unassignedCount: result.action_items.filter((a) => isUnspecified(a.owner)).length,
    noDueDateCount: result.action_items.filter((a) => isUnspecified(a.due_date)).length,
    riskCount: result.risks.length,
  };
}

File: backend/schemas/analysis.py
# =====================================================================
# backend/schemas/analysis.py
# 역할: API 요청/응답에서 사용하는 Pydantic 데이터 모델 정의
# - 프론트엔드의 src/types/analysis.ts 와 구조가 일치해야 함
# - 모든 Optional 필드는 AI가 원문에서 찾지 못할 때 None으로 반환
# =====================================================================

from pydantic import BaseModel
from typing import Optional


# ── 결정사항 ─────────────────────────────────────────────────────────
class Decision(BaseModel):
    """회의에서 내려진 결정사항 항목"""
    content: str                 # 결정 내용
    evidence_text: str           # 원문 근거 (AI가 발췌한 원본 텍스트)


# ── 실행 항목 ───────────────────────────────────────────────────────
class ActionItem(BaseModel):
    """회의에서 도출된 할 일 항목"""
    task: str                    # 해야 할 일
    owner: Optional[str]         # 담당자 (원문에 없으면 None → 프론트에서 '미정' 표시)
    due_date: Optional[str]      # 마감일 (원문에 없으면 None → 프론트에서 '미정' 표시)
    priority: Optional[str]      # 우선순위: "높음" | "보통" | "낮음" | None
    evidence_text: str           # 원문 근거


# ── 리스크 ────────────────────────────────────────────────────────────
class Risk(BaseModel):
    """회의에서 언급된 리스크 항목"""
    risk: str                    # 리스크 내용
    response: Optional[str]      # 대응 방향 (원문에 없으면 None)
    evidence_text: str           # 원문 근거


# ── 분석 결과 전체 ────────────────────────────────────────────────────
class AnalysisResult(BaseModel):
    """AI 분석의 최종 결과 구조 — 프론트의 AnalysisResult 타입과 동일"""
    summary: str                      # 전체 요약
    decisions: list[Decision]         # 결정사항 목록
    action_items: list[ActionItem]    # 실행 항목 목록
    risks: list[Risk]                 # 리스크 목록
    missing_info: list[str]           # 누락 정보 (문자열 목록)


# ── API 요청 모델 ─────────────────────────────────────────────────────
class AnalyzeTextRequest(BaseModel):
    """POST /api/analyze-text 요청 바디"""
    text: str                    # 분석할 회의록 또는 업무 메모 원문


class SaveGoogleDocRequest(BaseModel):
    """POST /api/save-google-doc 요청 바디"""
    title: str                   # 저장할 문서 이름 (빈 문자열이면 자동 생성)
    result: AnalysisResult       # 저장할 분석 결과


# ── API 응답 모델 ─────────────────────────────────────────────────────
class SaveGoogleDocResponse(BaseModel):
    """POST /api/save-google-doc 응답"""
    title: str                   # 실제로 저장된 문서명
    url: str                     # 생성된 Google Docs 문서 URL
